protocols/client/client.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2017 HyperHQ Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// gRPC client wrapper
+
+package client
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/mdlayher/vsock"
+	"google.golang.org/grpc"
+
+	agentgrpc "github.com/kata-project/agent/protocols/grpc"
+)
+
+const (
+	unixSocketScheme  = "unix"
+	vsockSocketScheme = "vsock"
+)
+
+type AgentClient struct {
+	agentgrpc.HyperstartServiceClient
+	conn *grpc.ClientConn
+}
+
+type dialer func(string, time.Duration) (net.Conn, error)
+
+// Supported sock address formats are:
+//   - unix://<unix socket path>
+//   - vsock://<cid>:<port>
+//   - <unix socket path>
+func NewAgentClient(sock string) (*AgentClient, error) {
+	addr, err := parse(sock)
+	if err != nil {
+		return nil, err
+	}
+	dialOpts := []grpc.DialOption{grpc.WithInsecure(), grpc.WithTimeout(5 * time.Second)}
+	dialOpts = append(dialOpts, grpc.WithDialer(agentDialer(addr)))
+	conn, err := grpc.Dial(sock, dialOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &AgentClient{
+		HyperstartServiceClient: agentgrpc.NewHyperstartServiceClient(conn),
+		conn: conn,
+	}, nil
+}
+
+func (c *AgentClient) Close() error {
+	return c.conn.Close()
+}
+
+func parse(sock string) (*url.URL, error) {
+	addr, err := url.Parse(sock)
+	if err != nil {
+		return nil, err
+	}
+
+	// validate more
+	switch addr.Scheme {
+	case vsockSocketScheme:
+		if addr.Hostname() == "" || addr.Port() == "" || addr.Path != "" {
+			return nil, errors.New("Invalid vsock scheme")
```

Comment:
Can you include `sock` in this error (and the two `errors.New()` that follow) to make debugging easier?